### ISSUE TITLE
Consuming external metrics with absolute values

### ISSUE THREAD

**Issue by svartalf** [metrics-rs/metrics#66]
It is very common to expose some external metrics as **counters** created outside of the current application, for example, by collecting system statistics (ex. total amount of bytes sent via the network interface) or fetching them from other application (as various Prometheus exporters do).
The problem is that these metrics has absolute values and with current API it is possible to increment counters only and there is no way to replace the value.

From my understanding, something like [Constant Metrics](https://godoc.org/github.com/prometheus/client_golang/prometheus#hdr-Custom_Collectors_and_constant_Metrics) can be added
to allow one-time loading of the metric values; this idea better to be discussed further.

Naive way would be to add `Recorder::set_counter` method, which change the value of the inner `AtomicU64`.
Obviously, it is not the best solution possible, yet, it can work as a temporary workaround till better idea will land the `metrics` crate; as an addition, it can be feature-gated with something like `#[cfg(feature = "unstable")]`.

Extra links:

 * https://github.com/tikv/rust-prometheus/pull/303
 * https://www.robustperception.io/setting-a-prometheus-counter
 * https://github.com/prometheus-net/prometheus-net/issues/73

**Comment by svartalf** [comment]
Follow-up after the discussion with @tobz, we talked about two possible ways to implement such a behaviour right now:

1. Skip the `metrics-runtime` crate at all and implement custom [`Recorder`](https://docs.rs/metrics/0.12.1/metrics/trait.Recorder.html), which will be able to set counter values directly.

While this is a reasonable option, I personally against it, as it would mean to fully re-implement `metrics-runtime`' `Recorder` from the scratch, and it might be too much for such a task

2. Check if it would be possible to use [`Sink::proxy`](https://docs.rs/metrics-runtime/0.12.1/metrics_runtime/struct.Sink.html#method.proxy) to create `Measurement`s.

This method is somewhat working, but has few quirks:

1. Creating proxy `Sink` with the empty name results in the underline char at the start of metric name, ex. `_bytes_sent_total 123456`
2. Closure that needs to be provided must be the `Fn() -> ..` type, which effectively means that it will be hard to pass any context into it
3. Since we are talking about **external** metrics, there is a chance that some kind of IO will be involved, and it is problematic to handle any errors in the proxy closure. Simple ignoring of any errors might not be the desired behavior in some cases.

In conclusion, while these methods could fit some cases greatly, there are caveats worth of discussing.

**Comment by tobz** [comment]
There's some renewed activity around this topic in #172 based on improvements we've made in `metrics` 0.13 that allow us to inject logic to the pipeline without having to necessarily extend the macros themselves.
